File: solutions/Strings/0020-valid-parentheses/valid-parentheses.cpp
class Solution {
public:

    bool isValid(string s) {
        stack<int> st;
        for(auto x:s){
            if(!st.empty()){
                char v=st.top();
                if((x==')' && v=='(') || (x=='}' && v=='{') || (x==']' && v=='[')){
                    st.pop();
                }else{
                    st.push(x);
                }
            }else{
                st.push(x);
            }
        }
        return st.empty();
    }

    bool isValid__basic_v1(string s) {
        unordered_map<char,char> mp;
        mp[')']='(';
        mp['}']='{';
        mp[']']='[';
        stack<int> st;
        for(auto x:s){
            if(!st.empty() && mp[x]==st.top()){
                st.pop();
            }else{
                st.push(x);
            }
        }
        return st.empty();
    }
};

File: solutions/Strings/0020-valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Deque<Character> st=new ArrayDeque<>();
        String open="({[";
        String closed=")}]";
        int n=s.length();
        for(int i=0;i<n;i++){
            if(open.indexOf(s.charAt(i))!=-1){
                st.push(s.charAt(i));
            }else{
                int index=closed.indexOf(s.charAt(i));
                if(!st.isEmpty() && open.indexOf(st.peek())==index){
                    st.pop();
                }else{
                    return false;
                }
            }
        }
        return st.isEmpty();
    }
}